### PR TITLE
Feature/dz 1 2

### PR DESCRIPTION
**Pull request by Admalll**
https://user-images.githubusercontent.com/80919249/199172034-05e6dcb2-3e2d-4b4b-b056-19cb20c5da1e.mp4

### CODE REVIEW

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
+// LoginViewController.swift
+// Copyright © RoadMap. All rights reserved.
+
+import UIKit
+
+/// Экран логина в приложение
+final class LoginViewController: UIViewController {
+    // MARK: - Constants
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:8]: Private Constants

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    // MARK: - Private IBOutlets
+
+    @IBOutlet private var scrollView: UIScrollView!
+    @IBOutlet private var appleLoginButton: UIButton!
+    @IBOutlet private var loginTextfield: UITextField!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupView()
+        subsribeOnKeyboardNotifications()
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:27]: You will not be resubscribed if your view is already loaded in memory, this should be in viewWillAppear

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        static let closeAlertButtonTitle = "Закрыть"
+    }
+
+    // MARK: - Private IBOutlets
+
+    @IBOutlet private var scrollView: UIScrollView!
+    @IBOutlet private var appleLoginButton: UIButton!
+    @IBOutlet private var loginTextfield: UITextField!
+
+    override func viewDidLoad() {
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:24]: // MARK: Lifecycle

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        scrollView.contentInset = contentInsets
+        scrollView.scrollIndicatorInsets = contentInsets
+    }
+
+    @objc private func handleKeyboardWillHideNotificationAction(_ notification: Notification) {
+        scrollView.contentInset = .zero
+        scrollView.scrollIndicatorInsets = .zero
+    }
+
+    @IBAction private func loginAction(_ sender: UIButton) {
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:93]: remove this method and put the login in `override func shouldPerformSegue`, if you do so, don't forget to guard check your segueIdentificator

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        let contentInsets = UIEdgeInsets(top: 0, left: 0, bottom: keyboardSize.height, right: 0)
+        scrollView.contentInset = contentInsets
+        scrollView.scrollIndicatorInsets = contentInsets
+    }
+
+    @objc private func handleKeyboardWillHideNotificationAction(_ notification: Notification) {
+        scrollView.contentInset = .zero
+        scrollView.scrollIndicatorInsets = .zero
+    }
+
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:92]: if you want to leave the method below - please add `//MARK: - Private IBActions` mark here

**Comment by Admalll** [comment]
забыл записать поведение скролла и клавиатуры

https://user-images.githubusercontent.com/80919249/199173666-c25e738e-40c6-40ec-b92b-9eee66dd5b8d.mp4

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    private func authenticateLoginData(loginData: String) {
+        guard loginData == Constants.correctAuthData else {
+            showLoginFailError(user: loginData)
+            return
+        }
+        performSegue(withIdentifier: Constants.tabBarIdentifier, sender: self)
+    }
+
+    private func showLoginFailError(user loginData: String) {
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:69]:
```suggestion
    private func showLoginFailedAlert(user loginData: String) {
```

Comment (valadzko) [reply RM_Vk/RM_Vk/LoginViewController.swift:69]: Fail and Error mean both the same and should not be used together, just pick one

File: RM_Vk/RM_Vk.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -9,7 +9,7 @@
 /* Begin PBXBuildFile section */
```
Comment (valadzko) [inline RM_Vk/RM_Vk.xcodeproj/project.pbxproj]: this file should not be committed

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    @objc private func handleKeyboardWillShowNotificationAction(_ notification: Notification) {
+        guard let info = notification.userInfo as? NSDictionary,
+              let keyboardSize = (info.value(forKey: UIResponder.keyboardFrameEndUserInfoKey) as? NSValue)?.cgRectValue
+              .size else { return }
+        let contentInsets = UIEdgeInsets(top: 0, left: 0, bottom: keyboardSize.height, right: 0)
+        scrollView.contentInset = contentInsets
+        scrollView.scrollIndicatorInsets = contentInsets
+    }
+
+    @objc private func handleKeyboardWillHideNotificationAction(_ notification: Notification) {
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:88]: probably "handle" in name can be removed without any damage to function meaning

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        )
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(handleKeyboardWillHideNotificationAction),
+            name: UIResponder.keyboardWillHideNotification,
+            object: nil
+        )
+    }
+
+    private func unsubscribeFromKeybordNotifications() {
```
Comment (valadzko) [inline RM_Vk/RM_Vk/LoginViewController.swift:56]: typo in Keyboard

**Review by valadzko — APPROVED** [review]

File: RM_Vk/RM_Vk/LoginViewController.swift (hunk 4 above)
Comment (Admalll) [reply RM_Vk/RM_Vk/LoginViewController.swift:93]: хочу оставить его, т.к. не хочу явно в сториборде переход делать

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+// Copyright © RoadMap. All rights reserved.
+
+import UIKit
+
+/// Экран логина в приложение
+final class LoginViewController: UIViewController {
+    // MARK: - Private constants
+
+    private enum Constants {
+        static let correctAuthData = "admin"
```
Comment (AlexanderKokh) [inline RM_Vk/RM_Vk/LoginViewController.swift:11]: Data?

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        static let loginFailAlertTitle = "Ошибка входа"
+        static let loginFailAlertMessage = "Не найден пользователь "
+        static let closeAlertButtonTitle = "Закрыть"
+    }
+
+    // MARK: - Private IBOutlets
+
+    @IBOutlet private var scrollView: UIScrollView!
+    @IBOutlet private var appleLoginButton: UIButton!
+    @IBOutlet private var loginTextfield: UITextField!
```
Comment (AlexanderKokh) [inline RM_Vk/RM_Vk/LoginViewController.swift:22]:
```suggestion
    @IBOutlet private let scrollView: UIScrollView!
    @IBOutlet private let appleLoginButton: UIButton!
    @IBOutlet private let loginTextfield: UITextField!
```

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        performSegue(withIdentifier: Constants.tabBarIdentifier, sender: self)
+    }
+
+    private func showLoginFail(user loginData: String) {
+        let alert = UIAlertController(
+            title: Constants.loginFailAlertTitle,
+            message: Constants.loginFailAlertMessage + loginData,
+            preferredStyle: .alert
+        )
+        alert.addAction(UIAlertAction(title: Constants.closeAlertButtonTitle, style: .default))
```
Comment (AlexanderKokh) [inline RM_Vk/RM_Vk/LoginViewController.swift:81]:
```suggestion
```
алерт создаем через экстеншен передавая параметры для его создания

File: RM_Vk/RM_Vk/LoginViewController.swift [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        scrollView.scrollIndicatorInsets = contentInsets
+    }
+
+    @objc private func keyboardWillHideNotificationAction(_ notification: Notification) {
+        scrollView.contentInset = .zero
+        scrollView.scrollIndicatorInsets = .zero
+    }
+
+    // MARK: - Private IBActions
+
```
Comment (AlexanderKokh) [inline RM_Vk/RM_Vk/LoginViewController.swift:100]: сначала IBActions потом Private methods

**Comment by Admalll** [comment]
https://user-images.githubusercontent.com/80919249/199447168-874949d2-8817-4beb-9b5e-44e16660955d.mp4

повороты

**Review by AlexanderKokh — APPROVED** [review]

File: RM_Vk/RM_Vk/LoginViewController.swift (hunk 11 above)
Comment (Admalll) [reply RM_Vk/RM_Vk/LoginViewController.swift:22]: let iboutlet`y нельзя поставить

**Pull request merged**